[lldb] Add count for number of DWO files loaded in statistics

qxy11 opened the pull request:
## Summary
A new `totalLoadedDwoFileCount` and `totalDwoFileCount` counters to available statisctics when calling "statistics dump".

1. `GetDwoFileCounts ` is created, and returns a pair of ints representing the number of loaded DWO files and the total number of DWO files, respectively. An override is implemented for `SymbolFileDWARF` that loops through each compile unit, and adds to a counter if it's a DWO unit, and then uses `GetDwoSymbolFile(false)` to check whether the DWO file was already loaded/parsed.

3. In `Statistics`, use `GetSeparateDebugInfo` to sum up the total number of loaded/parsed DWO files along with the total number of DWO files. This is done by checking whether the DWO file was already successfully `loaded` in the collected DWO data, anding adding to the `totalLoadedDwoFileCount`, and adding to `totalDwoFileCount` for all CU units.

## Expected Behavior
- When binaries are compiled with split-dwarf and separate DWO files, `totalLoadedDwoFileCount` would be the number of loaded DWO files and `totalDwoFileCount` would be the total count of DWO files.
- When using a DWP file instead of separate DWO files, `totalLoadedDwoFileCount` would be the number of parsed compile units, while `totalDwoFileCount` would be the total number of CUs in the DWP file. This should be similar to the counts we get from loading separate DWO files rather than only counting whether a single DWP file was loaded.
- When not using split-dwarf, we expect both `totalDwoFileCount` and `totalLoadedDwoFileCount` to be 0 since no separate debug info is loaded.

## Testing
**Manual Testing**
On an internal script that has many DWO files, `statistics dump` was called before and after a `type lookup` command. The `totalLoadedDwoFileCount` increased as expected after the `type lookup`.
```
(lldb) statistics dump
{
  ...
  "totalLoadedDwoFileCount": 29,
}
(lldb) type lookup folly::Optional<unsigned int>::Storage
typedef std::conditional<true, folly::Optional<unsigned int>::StorageTriviallyDestructible, folly::Optional<unsigned int>::StorageNonTriviallyDestructible>::type
typedef std::conditional<true, folly::Optional<unsigned int>::StorageTriviallyDestructible, folly::Optional<unsigned int>::StorageNonTriviallyDestructible>::type
...
(lldb) statistics dump
{
  ...
  "totalLoadedDwoFileCount": 2160,
}
```
**Unit test**
Added three unit tests that build with new "third.cpp" and "baz.cpp" files. For tests with w/ flags `-gsplit-dwarf -gpubnames`, this generates 2 DWO files. Then, the test incrementally adds breakpoints, and does a type lookup, and the count should increase for each of these as new DWO files get loaded to support these.
```
$ bin/lldb-dotest -p TestStats.py ~/llvm-sand/external/llvm-project/lldb/test/API/commands/statistics/basic/
----------------------------------------------------------------------
Ran 20 tests in 211.738s

OK (skipped=3)
```

qxy11 committed:
Add counter for number of DWO files loaded in statistics

qxy11 commented:
cc @dmpots @jeffreytan81 @clayborg

dmpots commented on lldb/include/lldb/Symbol/SymbolFile.h:
```diff
@@ -425,6 +425,11 @@ class SymbolFile : public PluginInterface {
+  /// Get the number of loaded DWO files for this symbol file.
+  /// This is used for statistics gathering and will return 0 for most
+  /// symbol file implementations except DWARF symbol files.
+  virtual uint32_t GetLoadedDwoFileCount() const { return 0; }
```
There is a `GetSeparateDebugInfo` member function that returns info about separate debug info (which dwo is one type). Looking [here](https://github.com/dmpots/llvm-project/blob/79108da325daec08f5b50169a9c35e03ea0645a3/lldb/source/Commands/CommandObjectTarget.cpp#L2650) it looks like we could use it to find all the dwos that were loaded.

I wonder if it would be better to use this existing function when generating the stats instead of adding a new overload here.

dmpots commented on lldb/test/API/commands/statistics/basic/TestStats.py:
```diff
@@ -512,6 +517,32 @@ def test_breakpoints(self):
+        various commands
+        """
+        da = {"CXX_SOURCES": "third.cpp baz.cpp", "EXE": self.getBuildArtifact("a.out")}
+        self.build(dictionary=da, debug_info=["dwo", "debug_names"])
```
Is building with debug names important for this test? I am trying to understand the interaction between debug names and dwo loading. For example, if we used `b baz` as the second command instead of `type lookup Baz` would the test produce the same results?

qxy11 replied:
Yeah, without the debug names, for the purpose of the test, all the DWO files get loaded when the target is created. If we used `b baz` as the second command instead of `type lookup Baz`, it'd still produce the same results.

qxy11 replied:
Sounds good, I added some more comments in the test about the debug names. From looking into the code, it seems like having the debug names table allows for the lazy loading w/ `DebugNamesDWARFIndex ` since we avoid calling the `Index` [method](https://github.com/llvm/llvm-project/blob/2ab9c35ea93f8557827d4cadcceb05e4eed2d30a/lldb/source/Plugins/SymbolFile/DWARF/ManualDWARFIndex.cpp#L34) in `ManualDWARFIndex`, and only trigger loading if `GetNonSkeletonUnit` gets [called](https://github.com/llvm/llvm-project/blob/2ab9c35ea93f8557827d4cadcceb05e4eed2d30a/lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp#L121).

dmpots commented on lldb/test/API/commands/statistics/basic/TestStats.py:
```diff
@@ -512,6 +517,32 @@ def test_breakpoints(self):
+        self.assertIn("totalLoadedDwoFileCount", debug_stats)
+        self.assertEqual(debug_stats["totalLoadedDwoFileCount"], 0)
+        
+        self.runCmd("b main")
```
I think it would be helpful to add some comments here about why we expect the count at each step. It makes it easier for someone to come in and understand what the test is trying to do if it ever fails in the future.

```
# Setting a breakpoint in main will load the main.dwo file so the count should be one.
...
self.assertEqual(...)

# Type lookup causes the second dwo file to be loaded.
...
```

jeffreytan81 commented:
Besides `totalLoadedDwoFileCount`, we need another `totalDwoFileCount`. The ratio between them will show us how much X% of dwo files have been loaded for the debug session. 

Also, we should test this against dwp file situation, and non-split-dwarf situation, what are the numbers we should report?

qxy11 committed:
Use GetSeparateDebugInfo and add totalLoadedDwoFileCount

Address comments to use `GetSeparateDebugInfo` instead of adding a new
Symbol File function. Also add a totalLoadedDwoFileCount to statistics
dump and add some unit tests for expected behavior in the DWP case  and
non-split-dwarf cases.

dmpots commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -353,6 +355,30 @@ llvm::json::Value DebuggerStats::ReportStatistics(
         for (const auto &symbol_module : symbol_modules.Modules())
           module_stat.symfile_modules.push_back((intptr_t)symbol_module.get());
       }
+      // Count DWO files from this symbol file using GetSeparateDebugInfo
```
Could we put this logic into a separate function like and call it here?

Something like
```
updateDwoCounts(sym_file, total_dwo_file_count, totoal_loaded_dwo_file_count)
```

qxy11 committed:
Put file counting logic in UpdateDwoFileCounts and add test_debug_names_eager_loads_dwo_files

Adds an extra unit test and refactors some logic into a helper function.
Also added additional comments in code and tests to clarify purpose of
debug_names usage and how it prevents the eager loading of DWO files

dmpots reviewed (approved):
LGTM! You should get approvals from @jeffreytan81 and @clayborg as well before merging.

clayborg commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -33,6 +33,37 @@ static void EmplaceSafeString(llvm::json::Object &obj, llvm::StringRef key,
+  StructuredData::Dictionary separate_debug_info;
+  if (sym_file->GetSeparateDebugInfo(separate_debug_info,
+                                     /*errors_only=*/false,
+                                     /*load_all_debug_info*/ false)) {
```
```
/*load_all_debug_info=*/false
```

clayborg commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -353,6 +386,8 @@ llvm::json::Value DebuggerStats::ReportStatistics(
         for (const auto &symbol_module : symbol_modules.Modules())
           module_stat.symfile_modules.push_back((intptr_t)symbol_module.get());
       }
+      UpdateDwoFileCounts(sym_file, total_dwo_file_count,
+                          total_loaded_dwo_file_count);
```
We should probably break this down and add statistics for each module using keys named "loadedDwoFileCount" and "dwoFileCount" in each module's statistics.  And then we still add totals at the end as you have. So maybe fetch into module_stat first and then update the totals:
```
UpdateDwoFileCounts(sym_file, module_stat.total_dwo_file_count,
                          module_stat.loaded_dwo_file_count);
total_dwo_file_count += module_stat.total_dwo_file_count;
total_loaded_dwo_file_count += module_stat.loaded_dwo_file_count;
```
We will need to add two things to ModuleStats:
```
uint32_t dwo_file_count = 0;
uint32_t loaded_dwo_file_count = 0;
```

qxy11 committed:
Add dwo_file_count and loaded_dwo_file_count to ModuleStats

jeffreytan81 commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -33,6 +33,37 @@ static void EmplaceSafeString(llvm::json::Object &obj, llvm::StringRef key,
+  // For DWP files, this increments counts for both total and successfully
+  // loaded DWO CUs. For non split-dwarf files, these counts should not change
+  StructuredData::Dictionary separate_debug_info;
+  if (sym_file->GetSeparateDebugInfo(separate_debug_info,
```
I would not recommend calling `GetSeparateDebugInfo` API for this task. `GetSeparateDebugInfo` API constructs `StructuredData::Dictionary` JSON style objects internally which is not cheap (I have seen profile traces that indicate constructing `StructuredData::Dictionary` as hot paths). This can get worse considering we are using `statistics dump` in lldb session logging by default. (means we are paying the cost every time).

If we want to reuse it, I would suggest refactoring and reuse the necessary underlying implementation without extra expensive burdens.

qxy11 committed:
Add a separate GetDwoFileCounts instead of using GetSeparateDebugInfo

jeffreytan81 commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.h:
```diff
@@ -282,6 +282,11 @@ class SymbolFileDWARF : public SymbolFileCommon {
                             bool errors_only) override;
 
+  // Gets a pair of loaded and total dwo file counts.
+  // For DWP files, this reports the counts for successfully loaded DWO CUs
```
"For DWP files", do you mean "For split dwarf files"? Otherwise, you need to mention "dwo files" in addition to "dwp files".

qxy11 committed:
Fix a typo in the comments

jeffreytan81 commented:
@vvereschaka, are you only seeing `TestStats.py` timeout on Windows or Linux as well? cc @qxy11 to follow-up.

jeffreytan81 commented:
> @jeffreytan81 ,
> 
> > @vvereschaka, are you only seeing `TestStats.py` timeout on Windows or Linux as well? cc @qxy11 to follow-up.
> 
> I seen it on Windows (non-cross) only. The linux builds get passed without that problem.

Thanks for checking. We did a relanding only enabling the test for dwo category, https://github.com/llvm/llvm-project/pull/145572. Let us know if there are still issues.
